db/kv_checksum.cc

```diff
@@ -0,0 +1,155 @@
+//  Copyright (c) 2020-present, Facebook, Inc.  All rights reserved.
+//  This source code is licensed under both the GPLv2 (found in the
+//  COPYING file in the root directory) and Apache 2.0 License
+//  (found in the LICENSE.Apache file in the root directory).
+
+#include "db/kv_checksum.h"
+
+#include <type_traits>
+
+namespace ROCKSDB_NAMESPACE {
+
+namespace {
+
+// bitwise operations on an enum class adapted from
+// https://softwareengineering.stackexchange.com/a/204566
+KvProtectionInfo::CoverageFlags operator|(KvProtectionInfo::CoverageFlags lhs,
+                                          KvProtectionInfo::CoverageFlags rhs) {
+  using T = std::underlying_type<KvProtectionInfo::CoverageFlags>::type;
+  return static_cast<KvProtectionInfo::CoverageFlags>(static_cast<T>(lhs) |
+                                                      static_cast<T>(rhs));
+}
+
+KvProtectionInfo::CoverageFlags& operator|=(
+    KvProtectionInfo::CoverageFlags& lhs, KvProtectionInfo::CoverageFlags rhs) {
+  lhs = lhs | rhs;
+  return lhs;
+}
+
+KvProtectionInfo::CoverageFlags operator&(KvProtectionInfo::CoverageFlags lhs,
+                                          KvProtectionInfo::CoverageFlags rhs) {
+  using T = std::underlying_type<KvProtectionInfo::CoverageFlags>::type;
+  return static_cast<KvProtectionInfo::CoverageFlags>(static_cast<T>(lhs) &
+                                                      static_cast<T>(rhs));
+}
+
+}  // anonymous namespace
+
+KvProtectionInfo::KvProtectionInfo() { Clear(); }
+
+Status KvProtectionInfo::VerifyAgainst(const KvProtectionInfo& expected) const {
+  if (expected.HasKeyChecksum() &&
+      GetKeyChecksum() != expected.GetKeyChecksum()) {
+    return Status::Corruption("key checksum mismatch");
+  }
+  if (expected.HasValueChecksum() &&
+      GetValueChecksum() != expected.GetValueChecksum()) {
+    return Status::Corruption("value checksum mismatch");
+  }
+  if (expected.HasSequenceNumber() &&
+      GetSequenceNumber() != expected.GetSequenceNumber()) {
+    return Status::Corruption("sequence number mismatch");
+  }
+  if (expected.HasColumnFamilyId() &&
+      GetColumnFamilyId() != expected.GetColumnFamilyId()) {
+    return Status::Corruption("column family ID mismatch");
+  }
+  if (expected.HasValueType() && GetValueType() != expected.GetValueType()) {
+    return Status::Corruption("value type mismatch");
+  }
```

Comment:
I believe this code will assert if the current flags do not contain all of the expected flags.  Was that the intent?  Or should that be a Corruption or InvalidArgument or something else?

Reply:
Yes, it is currently considered a programming error if we expect a field to be a particular value while that field is not present (but the other way around is a simple successful verification). One use case of this "superset" comparison is verification before we replace a field (like `ValueType` which is used differently in `WriteBatch` vs. `MemTable`; or checksum of value, which can be changed in merges after `max_successive_merges` is reached, or if an in-place update callback is invoked). Another use case is we have extra fields, like column family ID, which got added for `WriteBatch`, but aren't finally stored in `MemTable`.

A different design could be to provide APIs to replace/strip protection info for each field, while requiring the user to provide the expected old protection info so we can verify there was no corruption before replacing/removing it. If we use those correctly, the `KvProtectionInfo` comparison function can assume protection coverage is exactly the same.

Reply:
> so we can verify there was no corruption before replacing/removing it.

I'm not necessarily saying anything you don't know, but...

Verifying data often and with precise granularity is the enemy of performance, but not necessarily the enemy of data integrity. It is definitely the friend of debugability, but not necessarily appropriate for release builds.

For example, suppose in release builds instead of storing a hash for each of n fields, we just xor hashes (independent seeds! [0]) from all the n fields that have been set. Assuming your program logic can always reconstruct which inputs should have gone in, we can verify with essentially the same level of protection as the same size hash for each field.

With that in mind, one thing to consider is statically typing the set of inputs that have gone into a KvProtectionInfo. Although in theory there could be 2^n types to work with, hopefully we don't need to go outside a typical order very often. So using single-letter abbreviations, we could have

ProtectionInfoK
ProtectionInfoKV
ProtectionInfoKVS
ProtectionInfoKVSC
ProtectionInfoKVSCT

And each has functions (value semantics, const functions) to add the next level, back out one level (release mode check deferred until all levels are backed out), or replace any level from an old value to a new value (release mode check deferred). For full verification in release mode, you get to backing out the K from ProtectionInfoK, which should probably return Status.

[0] You can't really have independent seeds for the same kind of CRC checksum. Proof available upon request. That's why I'm going to say "hashes".